File: 201-300/202-happy-number.cpp
bool isHappy(int n) {
    int slow = n, fast = cal(n);
    while (fast != 1 && slow != 1 && fast != slow) {
        slow = cal(slow);
        fast = cal(cal(fast));
    }

    return (fast == 1 || slow == 1);
}

int cal(int n) {
    int tmp = 0;
    while (n >= 1) {
        tmp += (n % 10) * (n % 10);
        n = n / 10;
    }
    return tmp;
}

File: python version/202-happy-number.py
class Solution:
    def isHappy(self, n: int) -> bool:
        if n == 1: return True
        s = set()
        while n != 1 and n not in s:
            s.add(n)
            tmp = 0
            while n != 0:
                tmp += (n%10)**2
                n = n // 10
            n = tmp
        if n == 1:return True
        else: return False